Firestore/core/src/firebase/firestore/immutable/append_only_list.h

```diff
@@ -0,0 +1,228 @@
+ * that is not at the end, most operations will copy instead of chaining.
+ */
+template <typename T>
+class AppendOnlyList {
+ public:
+  using iterator = const T*;
+  using const_iterator = const T*;
+  using value_type = T;
+
+  AppendOnlyList() = default;
+
+  AppendOnlyList(std::initializer_list<T> initializer_list)
+      : contents_(std::make_shared<std::vector<T>>(initializer_list)),
+        size_(initializer_list.size()) {
+  }
+
+  /**
+   * Returns a new AppendOnlyList that has reserved the given capacity in its
+   * backing vector, without actually lengthening the chain.
+   *
+   * This has a similar effect to std::vector::reserve, except that *this is
+   * not actually modified. Successive `push_back` operations until `size()`
+   * is equal to `capacity` are guaranteed to be O(1).
+   *
+   * Note that if this instance is not the end of the chain then this forces
+   * a copy.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList reserve(size_t capacity) const {
+    if (capacity <= size_) {
+      return *this;
+    }
+
+    // Create a the underlying vector with capacity reserved, but return the
+    // result with the current size. Reserving does not actually append anything
+    // to the underlying vector so size() shouldn't change.
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(capacity);
+    return AppendOnlyList(std::move(new_contents), size_);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the given value appended to the end.
+   *
+   * Each `push_back` creates a new instance and appears not to modify any that
+   * come. If `push_back` is called on the last instance in a chain, it will
+   * share the backing vector with the prior instance.
+   *
+   * If `push_back` is called when this instance isn't the last instance in the
+   * chain, it will make a copy of all preceding elements in the chain and
+   * return a new chain suitable for further chained `push_back` operations.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList push_back(const T& value) const {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->push_back(value);
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the given value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList push_back(T&& value) const {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->push_back(std::move(value));
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList constructing a new value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  template <typename... Args>
+  ABSL_MUST_USE_RESULT AppendOnlyList emplace_back(Args&&... args) {
```

Comment:
It seems like this function should be `const`, for consistency with `push_back`?

Reply:
Yes. They all should be.